src/borrow_tracker/tree_borrows/mod.rs

```diff
@@ -285,43 +311,90 @@ trait EvalContextPrivExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
 
         let span = this.machine.current_span();
         let alloc_extra = this.get_alloc_extra(alloc_id)?;
-        let range = alloc_range(base_offset, ptr_size);
         let mut tree_borrows = alloc_extra.borrow_tracker_tb().borrow_mut();
 
-        // All reborrows incur a (possibly zero-sized) read access to the parent
-        if new_perm.initial_read {
-            tree_borrows.perform_access(
-                orig_tag,
-                Some((range, AccessKind::Read, diagnostics::AccessCause::Reborrow)),
-                this.machine.borrow_tracker.as_ref().unwrap(),
-                alloc_id,
-                this.machine.current_span(),
-            )?;
-        }
+        // Store initial permissions and their corresponding range.
+        let mut perms_map: RangeMap<Permission> = RangeMap::new(
+            ptr_size,
+            Permission::new_disabled(), // this will be overwritten
+        );
+        // Keep track of whether the node has any part that allows for interior mutability.
+        // FIXME: This misses `PhantomData<UnsafeCell<T>>` which could be considered a marker
+        // for requesting interior mutability.
+        let mut has_unsafe_cell = false;
+
+        // We should always do a read access to the `freeze` parts of the reference.
+        // This is important for properly updating the SIFA later:
+        // * If we do have an `UnsafeCell` (`has_unsafe_cell` becomes true), then
+        //   + some parts inside the initial range of the reference have `nonfreeze_perm`
+        //   + the "dynamic" area outside the initial range of the reference also have `nonfreeze_perm`
+        //   and thus it is sufficient to reset the strongest foreign access of all parents to the
+        //   `nonfreeze_perm` SIFA, because all the parts where we put `freeze_perm` already had a read
+        //   access happen, and thus the SFAs there are already reset. The assert below ensures that this
+        //   read access to the `freeze_perm` happens.
+        // * If there is no `UnsafeCell`, then `freeze_perm` is used everywhere (both inside and outside the initial range),
+        //   and we update everything to have the `freeze_perm`'s SIFA, so there are no issues. (And this assert below is not
+        //   actually needed in this case).
+        assert!(new_perm.freeze_access);
+
+        this.visit_freeze_sensitive(place, ptr_size, |range, frozen| {
+            has_unsafe_cell = has_unsafe_cell || !frozen;
+
+            // We are only ever `Frozen` inside the frozen bits.
+            let (perm_init, access) = if frozen {
+                (new_perm.freeze_perm, new_perm.freeze_access)
+            } else {
+                (new_perm.nonfreeze_perm, new_perm.nonfreeze_access)
+            };
+
+            // Store initial permissions.
+            for (_perm_range, perm) in perms_map.iter_mut(range.start, range.size) {
+                *perm = perm_init;
```

Comment:
here, it should create a new `LocationState` that is init if any only if `access` is set.